doc/man3/OSSL_CMP_exec_IR_ses.pod

```diff
@@ -0,0 +1,168 @@
+If no error occurred but no certificate is available yet then
+OSSL_CMP_try_certreq() remembers in the CMP context that polling should be retried
+and returns a positive 'checkAfter' value indicating the number of seconds
+the caller should let pass before trying again.
+In this case the caller can decide to sleep for some time and/or do anything else
+before retrying by calling OSSL_CMP_try_certreq() again with the same parameter
+values as before.
+In this case OSSL_CMP_try_certreq() polls to see
+whether meanwhile the requested certificate is available.
+In case the caller decides to abort the pending certificate request and provides
+a negative value as the B<req_type> argument then OSSL_CMP_try_certreq()
+aborts the CMP transaction by sending an error message to the server.
+
+OSSL_CMP_certConf_cb() is a basic certificate confirmation callback validating
+that the new certificate can be verified with the trusted/untrusted certificates
+in B<ctx>.
+As there is no requirement in RFC 4210 that the certificate can be
+validated by the client, this callback is not set by default in the context.
+
+OSSL_CMP_exec_RR_ses() requests the revocation of the certificate
+specified in the B<ctx> using L<OSSL_CMP_CTX_set1_oldCert(3)>.
+RFC 4210 is vague in which PKIStatus should be returned by the server.
+We take "accepted" and "grantedWithMods" as clear success and handle
+"revocationWarning" and "revocationNotification" just as warnings because CAs
+typically return them as an indication that the certificate was already revoked.
+"rejection" is a clear error. The values "waiting" and "keyUpdateWarning"
+make no sense for revocation and thus are treated as an error as well.
+
+OSSL_CMP_exec_GENM_ses() sends a general message containing the sequence of
+infoType and infoValue pairs (InfoTypeAndValue; short: B<ITAV>)
+provided in the B<ctx> using L<OSSL_CMP_CTX_push0_genm_ITAV(3)>.
+It returns the list of B<ITAV>s received in the GenRep.
+This can be used, for instance, to poll for CRLs or CA Key Updates.
+See RFC 4210 section 5.3.19 and appendix E.5 for details.
+
+=head1 NOTES
+
+CMP is defined in RFC 4210 (and CRMF in RFC 4211).
+
+So far the CMP client implementation is limited to one request per CMP message
+(and consequently to at most one response component per CMP message).
+
+=head1 RETURN VALUES
+
+OSSL_CMP_exec_IR_ses(), OSSL_CMP_exec_CR_ses(),
+OSSL_CMP_exec_P10CR_ses(), and OSSL_CMP_exec_KUR_ses() return a
+pointer to the newly obtained X509 certificate on success, B<NULL> on error.
+This pointer will be freed implicitly by OSSL_CMP_CTX_free() or
+CSSL_CMP_CTX_reinit().
+
+OSSL_CMP_try_certreq() returns -1 if the requested certificate is available
+via B<OSSL_CMP_CTX_get0_newCert()>, the checkAfter value (which is > 0)
+in case a 'waiting' status has been received (then B<OSSL_CMP_CTX_get0_newCert()>
+yields NULL),
+1 on successfully aborting a pending certificate request, and 0 on error.
```

Comment:
I don't particularly like overloading the return value to have all these different meanings. The certificate is available sounds like a success to me - so returning -1 to indicate it just seems odd and counter-intuitive. Perhaps better might be -1 for error, 0 for waiting and 1 for the certificate is available might be a more intuitive interface. In the event of "waiting" then there could be an additional function on the ctx to say how long is left until the application should poll again.

Reply:
Yeah, I agree that the various sorts of return values are a bit complicated.
I'll follow the -1 / 0 / +1 value you suggested, yet I'd like to a avoid adding a further function just for retrieving the checkAfter value - not only this would add one more function to the API but mainly because the meaning of the value could go wrong in case the app takes some time before retrieving it. Ok if I add instead an output parameter `int *checkAfter`?

Reply:
Yes - that would work too.

Reply:
Finished changing the result value structure.
It turns out that it is more straightforward to use - as usual - 1 for success and 0 for error, 
while -1 indicates the 'waiting' condition (in which case `*checkAfter` is assigned unless `checkAfter == NULL`).